core/src/main/java/org/apache/accumulo/core/clientImpl/OfflineTableExtentCache.java

```diff
@@ -0,0 +1,170 @@
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Scheduler;
+
+public class OfflineTableExtentCache {
+
+  private static final Logger LOG = LoggerFactory.getLogger(OfflineTableExtentCache.class);
+
+  private final Cache<TableId,TreeMap<Range,List<KeyExtent>>> cache;
+  private final ClientContext ctx;
+
+  OfflineTableExtentCache(ClientContext ctx) {
+    this.ctx = ctx;
+    cache = Caffeine.newBuilder().scheduler(Scheduler.systemScheduler())
+        .expireAfterAccess(Duration.ofMinutes(10)).build();
+  }
+
+  private List<KeyExtent> loadExtentsForRange(TableId tid, Range r) {
+    LOG.trace("Loading extents for tid: {} and range: {}", tid, r);
+    TabletsMetadata tm =
+        ctx.getAmple().readTablets().forTable(tid)
+            .overlapping(r.getStartKey() == null ? null : r.getStartKey().getRow(),
+                r.isStartKeyInclusive(), r.getEndKey() == null ? null : r.getEndKey().getRow())
+            .build();
+    List<KeyExtent> result = new ArrayList<>();
+    for (TabletMetadata t : tm) {
+      LOG.trace("Evaluating: {}", t.getExtent());
+      if (t.getExtent().endRow() == null || r.getEndKey() == null || t.getExtent().endRow() != null
+          && r.getEndKey().compareTo(new Key(t.getExtent().endRow()), PartialKey.ROW) >= 0) {
+        LOG.trace("Adding: {}", t.getExtent());
+        result.add(t.getExtent());
+      }
+    }
+    LOG.trace("Loaded extents: {}", result);
+    return result;
+  }
+
+  public Map<Range,List<KeyExtent>> lookup(TableId tid, List<Range> ranges) {
+    LOG.trace("Performing lookup for table: {}, range: {}", tid, ranges);
+    Map<Range,List<KeyExtent>> results = new TreeMap<>();
+    TreeMap<Range,List<KeyExtent>> extentMap = cache.get(tid, (t) -> {
+      return new TreeMap<Range,List<KeyExtent>>();
+    });
+
+    Collections.sort(ranges);
+    ranges.forEach(r -> {
+      // Find an existing range in the extentMap that overlaps the scan range
+      boolean foundRange = false;
+      Range rangeBefore = extentMap.floorKey(r);
+      if (rangeBefore != null) {
+        Map<Range,List<KeyExtent>> tm = extentMap.tailMap(rangeBefore);
+        for (Entry<Range,List<KeyExtent>> e : tm.entrySet()) {
+          if (e.getKey().overlaps(r)) {
+            // Ignore extents that don't contain the scan range
+            List<KeyExtent> extents = new ArrayList<>();
+            if (r.getStartKey() == null) {
+              extents.addAll(e.getValue());
+            } else {
+              e.getValue().forEach(extent -> {
+                Text row = r.isStartKeyInclusive() ? r.getStartKey().getRow()
+                    : r.getStartKey().followingKey(PartialKey.ROW).getRow();
+                if (extent.contains(row)) {
+                  extents.add(extent);
+                }
+              });
+            }
+            results.put(r, extents);
+            foundRange = true;
+            break;
```

Comment:
Getting a bit lost w/ the curly braces, but it seems like this takes only the first range in the cache that overlaps the query range.  If so, what if there are other ranges in the cache that overlap the query range?